### PR TITLE
[feat]: Attn-QAT inference + training backends (deadcode) (Attn-QAT 4/12)

### PR DESCRIPTION
**Pull request by SolitaryThinker**
Slice 4 of 12 in the PR #1225 decomposition. Deadcode — adds the two QAT attention backends but does NOT register them in the selector. Activation happens at slice 12.

## What this adds

- `fastvideo/attention/backends/attn_qat_infer.py` (+121)
- `fastvideo/attention/backends/attn_qat_train.py` (+145)

Both files extracted from PR #1225 (`3f818d0fc532ec6494b465967d5f485150917d0c`), audited for newest-API compat against `origin/main`, and retargeted from `fastvideo/training/` to `fastvideo/train/` where applicable per Will's directive.

## Out of scope (in later slices)

- Selector registration → slice 12 (activation)
- Shared attention infra additions for QAT-compat → slice 5
- Linear/MLP FP4 path additions → slice 6

## Stacked on

This PR is stacked on top of PR #1350 (slice 3, FP4 linear). When PR #1350 merges, GitHub auto-retargets this PR's base to `main`; the diff shown will be just slice 4's delta.

Attn-QAT-Stack: 4/12

Co-Authored-By: jzhang38 <42993249+jzhang38@users.noreply.github.com>
Co-Authored-By: RandNMR73 <99706358+RandNMR73@users.noreply.github.com>

### CODE REVIEW

File: fastvideo/attention/backends/attn_qat_train.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    except ImportError:
+        _attn_qat_train_attention = None
+
+    return _attn_qat_train_attention
+
+
+def attn_qat_train(q_BLHD: torch.Tensor,
+                   k_BLHD: torch.Tensor,
+                   v_BLHD: torch.Tensor,
+                   is_causal: bool = False) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/attention/backends/attn_qat_train.py:55]: Done in e0b87964. Added optional sm_scale param to attn_qat_train (defaults to None → 1/sqrt(d) fallback).

File: fastvideo/attention/backends/attn_qat_train.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        self.dropout = extra_impl_args.get("dropout_p", 0.0)
+
+    def forward(
+        self,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_metadata: AttentionMetadata,
+    ) -> torch.Tensor:
+        return attn_qat_train(query, key, value, is_causal=self.causal)
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/attention/backends/attn_qat_train.py:145]: Done in e0b87964. Plumbed self.softmax_scale through forward() to attn_qat_train.

File: fastvideo/attention/backends/attn_qat_infer.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        key = key.transpose(1, 2).contiguous()
+        value = value.transpose(1, 2).contiguous()
+
+        output = attn_qat_infer(
+            query,
+            key,
+            value,
+            attn_mask=None,
+            is_causal=self.causal,
+        )
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/attention/backends/attn_qat_infer.py:120]: Done in e0b87964. Passed sm_scale=self.softmax_scale to the kernel call.

File: fastvideo/attention/backends/attn_qat_train.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    use_qat_qkv_backward = True
+    smooth_k = False
+    warp_specialize = True
+    is_qat = True
+    two_level_quant_p_sage3 = False
+    fake_quant_p_bwd = True
+    use_high_prec_o = True
+    smooth_q = False
+    sm_scale = 1.0 / (q_BHLD.shape[-1]**0.5)
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/attention/backends/attn_qat_train.py:73]: Done in e0b87964. Uses passed sm_scale; falls back to 1/sqrt(d) when None.

File: fastvideo/attention/backends/attn_qat_infer.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        head_size: int,
+        causal: bool,
+        softmax_scale: float,
+        num_kv_heads: int | None = None,
+        prefix: str = "",
+        **extra_impl_args,
+    ) -> None:
+        self.causal = causal
+        self.softmax_scale = softmax_scale
+        self.dropout = extra_impl_args.get("dropout_p", 0.0)
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/attention/backends/attn_qat_infer.py:96]: Done in e0b87964. Raises NotImplementedError if dropout_p > 0 (kernel does not support dropout).

File: fastvideo/attention/backends/attn_qat_train.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        head_size: int,
+        causal: bool,
+        softmax_scale: float,
+        num_kv_heads: int | None = None,
+        prefix: str = "",
+        **extra_impl_args,
+    ) -> None:
+        self.causal = causal
+        self.softmax_scale = softmax_scale
+        self.dropout = extra_impl_args.get("dropout_p", 0.0)
```
[the remark this answers is not in the document]
Comment (SolitaryThinker) [reply fastvideo/attention/backends/attn_qat_train.py:136]: Done in e0b87964. Raises NotImplementedError if dropout_p > 0 (training kernel does not support dropout).

**Comment by SolitaryThinker** [comment]
This PR was auto-closed when its base branch `pr1225_s3` was deleted upstream after PR #1350 (slice 3) merged. Replacement PR opened at #1358 with the original 4 commits + 6 review-fix commits squashed into one — gemini-bot's 6 inline comments addressed there (replies posted on each thread above for reference).
